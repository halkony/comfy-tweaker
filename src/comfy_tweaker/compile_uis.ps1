pyside6-uic .\ui\tweaker.ui -o ./ui/main_ui.py
pyside6-uic .\ui\preferences.ui -o ./ui/preferences_ui.py
pyside6-uic .\ui\supporters.ui -o ./ui/supporters_ui.py