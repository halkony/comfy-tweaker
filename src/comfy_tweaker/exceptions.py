class InvalidSelectorError(Exception):
    pass


class NonUniqueSelectorError(Exception):
    pass


class IncompleteImageWorkflowError(Exception):
    pass


class NodeNotFoundError(Exception):
    pass


class NodeFieldNotFound(Exception):
    pass
