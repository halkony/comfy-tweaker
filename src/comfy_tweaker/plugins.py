from dataclasses import dataclass, field
from enum import Enum
import importlib.util
import sys

def import_plugin(module_name, plugin_path):
    spec = importlib.util.spec_from_file_location(module_name, plugin_path)
    plugin = importlib.util.module_from_spec(spec)
    sys.modules[module_name] = plugin
    spec.loader.exec_module(plugin)
    return plugin


class PluginType(Enum):
    GLOBALS = "globals"
    FILTERS = "filters"


@dataclass
class Plugin:
    name: str
    func: callable
    plugin_type: PluginType = field(default=PluginType.GLOBALS.value)
