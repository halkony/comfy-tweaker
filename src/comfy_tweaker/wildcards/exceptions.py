class WildcardNotFound(Exception):
    pass


class InvalidWildcardFormat(Exception):
    pass


class EmptyWildcardFile(Exception):
    pass
