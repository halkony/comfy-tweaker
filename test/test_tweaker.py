import asyncio
import os

import pytest

import comfy_tweaker as tweaker
from comfy_tweaker import Tweak, Tweaks, Workflow
from comfy_tweaker.exceptions import (IncompleteImageWorkflowError,
                                      InvalidSelectorError)


@pytest.fixture
def workflow(tweaks_directory):
    return Workflow.from_image(tweaks_directory / "valid_workflow_image.png")


@pytest.fixture
def tweaks(tweaks_directory):
    return Tweaks.from_file(tweaks_directory / "tweaks_file.yaml")


def test_tweak_must_have_valid_selector():
    with pytest.raises(InvalidSelectorError):
        tweak = tweaker.Tweak(
            {"nonsense": "nonsense"}, {"lora_name": "test.safetensors"}
        )


def test_tweak_cant_have_multiple_selectors():
    with pytest.raises(InvalidSelectorError):
        tweak = tweaker.Tweak(
            {"id": "346", "name": "test"}, {"lora_name": "test.safetensors"}
        )


def test_can_define_tweak():
    """Tweaks are single changes applied to a node"""
    tweak = tweaker.Tweak({"id": "346"}, {"lora_name": "test.safetensors"})
    assert tweak is not None


def test_tweak_is_immutable():
    tweak = tweaker.Tweak({"id": "346"}, {"lora_name": "test.safetensors"})
    with pytest.raises(AttributeError):
        tweak.selector = {"name": "test"}


def test_can_create_tweaks():
    """Tweaks are collections of tweaks that can be applied to a workflow"""
    tweaks = tweaker.Tweaks()
    assert tweaks is not None


def test_can_validate_tweaks(workflow, tweaks):
    workflow.validate(tweaks)


def test_workflow_from_image_raises_error_with_invalid_image(tweaks_directory):
    with pytest.raises(IncompleteImageWorkflowError):
        workflow = tweaker.Workflow.from_image(
            tweaks_directory / "invalid_workflow_image.png"
        )


def test_can_get_tweaks_from_file(snapshot, tweaks_directory):
    tweaks_file_path = tweaks_directory / "tweaks_file.yaml"

    tweaks = tweaker.Tweaks.from_file(tweaks_file_path)
    assert tweaks is not None
    assert tweaks.tweaks == snapshot


def test_can_create_workflow_from_image(tweaks_directory, snapshot):
    workflow = tweaker.Workflow.from_image(tweaks_directory / "valid_workflow_image.png")
    assert workflow == snapshot


def test_creating_workflow_with_invalid_image_raises_error(tweaks_directory):
    with pytest.raises(IncompleteImageWorkflowError):
        workflow = tweaker.Workflow.from_image(
            tweaks_directory / "invalid_workflow_image.png"
        )

def test_can_save_workflow_to_files(tweaks_directory, snapshot):
    workflow = tweaker.Workflow.from_image(tweaks_directory / "valid_workflow_image.png")
    new_gui_workflow_path = tweaks_directory / "new_example_workflow.json"
    new_api_workflow_path = tweaks_directory / "new_API_example_workflow.json"

    workflow.save(new_gui_workflow_path, new_api_workflow_path)
    assert new_gui_workflow_path.exists()
    assert new_api_workflow_path.exists()


def test_can_save_tweaks_to_yaml_file(snapshot, tmpdir):
    tweaks_file_path = tmpdir / "tweaks_file.yaml"
    tweaks = Tweaks()
    tweaks = tweaks.add(tweaker.Tweak({"id": "346"}, {"lora_name": "test.safetensors"}))

    tweaks.save(tweaks_file_path)
    tweaks = Tweaks.from_file(tweaks_file_path)
    assert tweaks == snapshot


def test_models_can_take_match_arg(models_directory):
    os.environ["MODELS_FOLDER"] = str(models_directory)
    for _ in range(200):
        tweaks_yaml = """
        tweaks:
            - selector:
                id: 346
              changes:
                lora_name: {{ from_models_folder("lora", match="1") }}
        """
        tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
        assert tweaks.tweaks[0].changes["lora_name"] == "test1.safetensors"


def test_models_can_take_regex_match_arg(models_directory):
    os.environ["MODELS_FOLDER"] = str(models_directory)
    for _ in range(200):
        tweaks_yaml = """
        tweaks:
            - selector:
                id: 346
              changes:
                lora_name: {{ from_models_folder("lora", regex_match=".*?1\.safetensors") }}
        """
        tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
        assert tweaks.tweaks[0].changes["lora_name"] == "test1.safetensors"


def test_can_get_random_choice(mocker):
    tweaks_yaml = """
    tweaks:
        - selector:
            id: 346
          changes:
            lora_name: {{ random_choice(["test.safetensors", "test.safetensors2"]) }}
    """
    expected_value = "expected_value.safetensors"
    mocked_random_choice = mocker.patch("random.choice", return_value=expected_value)

    tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
    mocked_random_choice.assert_called_with(["test.safetensors", "test.safetensors2"])
    assert tweaks.tweaks[0].changes["lora_name"] == expected_value


def test_can_evaluate_rng_tweak_tags(tweaks_directory, mocker):
    yaml_string = """
    tweaks:
        - selector:
            id: 346
          changes:
            strength: {{ random_float(0.0, 1.0) }}
    """
    mocked_random = mocker.patch("random.uniform", return_value=0.5)
    tweaks = tweaker.Tweaks.from_yaml(yaml_string)

    assert tweaks.tweaks[0].changes["strength"] == 0.5
    mocked_random.assert_called_with(0.0, 1.0)


def test_can_apply_tweaks_to_workflow(tweaks_directory, snapshot):
    workflow_image_path = tweaks_directory / "valid_workflow_image.png"
    tweaks_file_path = tweaks_directory / "tweaks_file.yaml"

    tweaks = tweaker.Tweaks.from_file(tweaks_file_path)
    workflow = tweaker.Workflow.from_image(workflow_image_path)

    result = workflow.apply_tweaks(tweaks)
    assert result == snapshot


def test_can_use_wildcards_in_tweaks_file(mocker):
    tweaks_data = """
    tweaks:
        - selector:
            id: 346
          changes:
            value: {{  "__animals__" | wildcards }}
    """

    expected_result = "This is expected."
    mocker.patch(
        "comfy_tweaker.wildcards.WildcardProcessor.process", return_value=expected_result
    )

    tweaks = tweaker.Tweaks.from_yaml(tweaks_data)
    assert tweaks.tweaks[0].changes == {"value": expected_result}


def test_tweaks_file_wildcards_can_use_file_wildcards(mocker, wildcards_directory):
    tweaks_data = """
    tweaks:
        - selector:
            id: 346
          changes:
            value: {{ "{__animals__}" | wildcards }}
    """
    os.environ["WILDCARDS_DIRECTORY"] = str(wildcards_directory)

    tweaks = tweaker.Tweaks.from_yaml(tweaks_data)
    assert tweaks.tweaks[0].changes["value"] in ["cow", "dog", "cat", "moose"]


def test_tweaks_file_can_product_random_int(mocker):
    mocked_randint = mocker.patch("random.randint", return_value=50)

    tweaks_data = """
    tweaks:
        - selector:
            id: 346
          changes:
            strength: {{random_int(0, 100)}}
    """
    tweaks = tweaker.Tweaks.from_yaml(tweaks_data)
    mocked_randint.assert_called_with(0, 100)
    assert tweaks.tweaks[0].changes == {"strength": 50}


def test_can_get_models_from_folder(mocker, models_directory):
    os.environ["MODELS_FOLDER"] = str(models_directory)

    tweaks_yaml = """
    tweaks:
        - selector:
            id: 346
          changes:
            lora_name: {{ from_models_folder("lora") }}
    """
    tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
    assert tweaks.tweaks[0].changes["lora_name"] in [
        "test1.safetensors",
        "test2.safetensors",
    ]


def test_can_match_to_lora_names(models_directory):
    os.environ["MODELS_FOLDER"] = str(models_directory)

    # this isnt how you would use it practically, but returning the list directly keeps rng out of our test
    tweaks_yaml = """
    tweaks:
        - selector:
            id: 346
          changes:
            lora_name: {{ from_models_folder("lora", match="1") }}
    """
    tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
    assert tweaks.tweaks[0].changes["lora_name"] == "test1.safetensors"


def test_tweaks_id_selector_works_with_ints_and_string():
    tweaks_yaml = """
    tweaks:
        - selector:
            id: 346
          changes:
            strength: 50
        - selector:
            id: "348"
          changes:
            strength: 50
    """
    tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
    assert tweaks.tweaks[0].selector["id"] == "346"
    assert tweaks.tweaks[1].selector["id"] == "348"


def test_can_reusue_random_values_with_jinja():
    tweaks_yaml = """
    {% set random_value = random_int(0, 100) %}
    tweaks:
        - selector:
            id: 346
          changes:
            strength: {{ random_value }}
        - selector:
            id: 348
          changes:
            strength: {{ random_value }}
    """
    tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
    assert tweaks.tweaks[0].changes["strength"] == tweaks.tweaks[1].changes["strength"]


def test_tweaks_file_can_cycle_through_files_in_folder(tmpdir):
    tweaks_yaml = f"""
    tweaks:
        - selector:
            id: 346
          changes:
            lora_name: {{{{ from_file_in_folder("{str(tmpdir).replace(os.sep, "/")}", cycle=True) }}}}
    """
    file1 = tmpdir / "file1.txt"
    file2 = tmpdir / "file2.txt"

    file1.write_text("file1", encoding="utf-8")
    file2.write_text("file2", encoding="utf-8")

    tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
    assert tweaks.tweaks[0].changes["lora_name"] == "file1"
    for i in range(200):
        tweaks = tweaks.regenerate()
        if i % 2 == 0:
            expected_value = "file2"
        else:
            expected_value = "file1"
        assert tweaks.tweaks[0].changes["lora_name"] == expected_value


def test_tweaks_file_supports_folder_cycle(models_directory):
    os.environ["MODELS_FOLDER"] = str(models_directory)

    tweaks_yaml = """
    tweaks:
        - selector:
            id: 346
          changes:
            lora_name: {{ from_models_folder("lora", cycle=True) }}
    """

    tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
    assert tweaks.tweaks[0].changes["lora_name"] == "test1.safetensors"
    for i in range(200):
        tweaks = tweaks.regenerate()
        if i % 2 == 0:
            expected_value = "test2.safetensors"
        else:
            expected_value = "test1.safetensors"
        assert tweaks.tweaks[0].changes["lora_name"] == expected_value


def test_empty_tweaks_wont_change_workflow(workflow):
    tweaks = Tweaks()
    result = workflow.apply_tweaks(tweaks)
    assert result == workflow


def test_can_regex_match_to_lora_names(models_directory):
    os.environ["MODELS_FOLDER"] = str(models_directory)

    # this isnt how you would use it practically, but returning the list directly keeps rng out of our test
    tweaks_yaml = """
    tweaks:
        - selector:
            id: 346
          changes:
            lora_name: {{ in_models_folder("lora") | regex_match(".*?2.*?") }}
        - selector:
            id: 348
          changes:
            lora_name: {{ in_models_folder("lora") }}
    """
    tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
    assert tweaks.tweaks[0].changes["lora_name"] == ["test2.safetensors"]
    assert tweaks.tweaks[1].changes["lora_name"] == [
        "test1.safetensors",
        "test2.safetensors",
    ]

def test_can_make_xy_matrix_in_jinja():
    tweaks_yaml = """
    {% macro lora1_strength(iteration) %}{{ (iteration // 10) * 0.05 }}{% endmacro %}
    {% macro lora2_strength(iteration) %}{{ (iteration % 10) * 0.05 }}{% endmacro %}
    tweaks:
        - selector:
            name: "Lora Loader 1"
          changes:
            strength: {{ lora1_strength(iteration) }}
        - selector:
            name: "Lora Loader 2"
          changes:
            strength: {{ lora2_strength(iteration) }}
    """

    tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
    for i in range(200):
        assert tweaks.tweaks[0].changes["strength"] == (i // 10) * 0.05
        assert tweaks.tweaks[1].changes["strength"] == (i % 10) * 0.05
        tweaks = tweaks.regenerate()

def test_can_generate_tweaks_with_an_iterator():
    tweaks_yaml = """
    tweaks:
        - selector:
            id: 346
          changes:
            lora_strength: {{ iteration }}
    """
    tweaks = tweaker.Tweaks.from_yaml(tweaks_yaml)
    for i in range(200):
        assert tweaks.tweaks[0].changes["lora_strength"] == i
        tweaks = tweaks.regenerate()

@pytest.mark.skip("Mocker not functional")
@pytest.mark.asyncio
async def test_job_queue_start(workflow, tweaks, mocker):
    mocked_comfyui = mocker.patch("comfy_tweaker.comfyui.send_workflow_to_server")
    job = tweaker.Job(workflow, tweaks, amount=2)
    queue = tweaker.JobQueue(queue=[job])
    task = asyncio.create_task(queue.start())

    await task
    assert len(queue.queue) == 0
    assert mocked_comfyui.send_workflow_to_server.call_count == 2


@pytest.mark.skip("Mocker not functional")
@pytest.mark.asyncio
async def test_job_queue_pause(workflow, tweaks, mock_send_workflow_to_server):
    job = tweaker.Job(workflow, tweaks, amount=10)
    queue = tweaker.JobQueue(queue=[job])
    task = asyncio.create_task(queue.start())
    raise AssertionError("Queue not processing jobs")
    attempts = 15
    while mock_send_workflow_to_server.call_count < 5:
        if attempts == 0:
            raise AssertionError("Queue not processing jobs")
        attempts -= 1

    queue.pause()
    await task
    assert len(queue.queue) == 1
    assert mock_send_workflow_to_server.call_count == 5
