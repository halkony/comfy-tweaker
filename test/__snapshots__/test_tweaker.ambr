# serializer version: 1
# name: test_can_apply_tweaks_to_workflow
  Workflow(gui_workflow={'last_node_id': 8, 'last_link_id': 14, 'nodes': [{'id': 1, 'type': 'CheckpointLoaderSimple', 'pos': [593.9912109375, 1441.2381591796875], 'size': [315, 98], 'flags': {}, 'order': 1, 'mode': 0, 'inputs': [], 'outputs': [{'name': 'MODEL', 'type': 'MODEL', 'links': [10], 'slot_index': 0}, {'name': 'CLIP', 'type': 'CLIP', 'links': [12], 'slot_index': 1}, {'name': 'VAE', 'type': 'VAE', 'links': [6], 'slot_index': 2}], 'properties': {'Node name for S&R': 'CheckpointLoaderSimple'}, 'widgets_values': ['juggernautXL_juggernautX.safetensors']}, {'id': 2, 'type': 'KSampler', 'pos': [2061.75390625, 1429.724365234375], 'size': [315, 474], 'flags': {}, 'order': 5, 'mode': 0, 'inputs': [{'name': 'model', 'type': 'MODEL', 'link': 11}, {'name': 'positive', 'type': 'CONDITIONING', 'link': 2}, {'name': 'negative', 'type': 'CONDITIONING', 'link': 3}, {'name': 'latent_image', 'type': 'LATENT', 'link': 7}], 'outputs': [{'name': 'LATENT', 'type': 'LATENT', 'links': [8], 'slot_index': 0}], 'properties': {'Node name for S&R': 'KSampler'}, 'widgets_values': [316381668905581, 'randomize', 20, 8, 'euler', 'normal', 1]}, {'id': 3, 'type': 'CLIPTextEncode', 'pos': [1592.681396484375, 1540.5108642578125], 'size': [400, 200], 'flags': {}, 'order': 3, 'mode': 0, 'inputs': [{'name': 'clip', 'type': 'CLIP', 'link': 13}], 'outputs': [{'name': 'CONDITIONING', 'type': 'CONDITIONING', 'links': [2]}], 'properties': {'Node name for S&R': 'CLIPTextEncode'}, 'widgets_values': ['An old man on a park bench']}, {'id': 4, 'type': 'CLIPTextEncode', 'pos': [1596.7044677734375, 1803.4591064453125], 'size': [400, 200], 'flags': {}, 'order': 4, 'mode': 0, 'inputs': [{'name': 'clip', 'type': 'CLIP', 'link': 14}], 'outputs': [{'name': 'CONDITIONING', 'type': 'CONDITIONING', 'links': [3]}], 'properties': {'Node name for S&R': 'CLIPTextEncode'}, 'widgets_values': ['']}, {'id': 5, 'type': 'VAEDecode', 'pos': [2473.87451171875, 1391.7969970703125], 'size': [210, 46], 'flags': {}, 'order': 6, 'mode': 0, 'inputs': [{'name': 'samples', 'type': 'LATENT', 'link': 8}, {'name': 'vae', 'type': 'VAE', 'link': 6}], 'outputs': [{'name': 'IMAGE', 'type': 'IMAGE', 'links': [9], 'slot_index': 0}], 'properties': {'Node name for S&R': 'VAEDecode'}}, {'id': 6, 'type': 'EmptyLatentImage', 'pos': [2057.55859375, 1739.7811279296875], 'size': [315, 106], 'flags': {}, 'order': 0, 'mode': 0, 'inputs': [], 'outputs': [{'name': 'LATENT', 'type': 'LATENT', 'links': [7]}], 'properties': {'Node name for S&R': 'EmptyLatentImage'}, 'widgets_values': [1024, 1024, 1]}, {'id': 7, 'type': 'SaveImage', 'pos': [2466.687744140625, 1545.7578125], 'size': [315, 270], 'flags': {}, 'order': 7, 'mode': 0, 'inputs': [{'name': 'images', 'type': 'IMAGE', 'link': 9}], 'outputs': [], 'properties': {}, 'widgets_values': ['ComfyUI']}, {'id': 8, 'type': 'LoraLoader', 'pos': [1196.7196044921875, 1665.0206298828125], 'size': [315, 126], 'flags': {}, 'order': 2, 'mode': 0, 'inputs': [{'name': 'model', 'type': 'MODEL', 'link': 10}, {'name': 'clip', 'type': 'CLIP', 'link': 12}], 'outputs': [{'name': 'MODEL', 'type': 'MODEL', 'links': [11], 'slot_index': 0}, {'name': 'CLIP', 'type': 'CLIP', 'links': [13, 14], 'slot_index': 1}], 'properties': {'Node name for S&R': 'LoraLoader'}, 'widgets_values': ['Hyper-SDXL-12steps-CFG-lora.safetensors', 1, 1]}], 'links': [[2, 3, 0, 2, 1, 'CONDITIONING'], [3, 4, 0, 2, 2, 'CONDITIONING'], [6, 1, 2, 5, 1, 'VAE'], [7, 6, 0, 2, 3, 'LATENT'], [8, 2, 0, 5, 0, 'LATENT'], [9, 5, 0, 7, 0, 'IMAGE'], [10, 1, 0, 8, 0, 'MODEL'], [11, 8, 0, 2, 0, 'MODEL'], [12, 1, 1, 8, 1, 'CLIP'], [13, 8, 1, 3, 0, 'CLIP'], [14, 8, 1, 4, 0, 'CLIP']], 'groups': [], 'config': {}, 'extra': {'ds': {'scale': 1.6940000000000053, 'offset': [-1357.8089102097, -1514.3590659107385]}, 'node_versions': {'comfy-core': '0.3.12'}, 'VHS_latentpreview': False, 'VHS_latentpreviewrate': 0}, 'version': 0.4}, api_workflow={'1': {'inputs': {'ckpt_name': 'juggernautXL_juggernautX.safetensors'}, 'class_type': 'CheckpointLoaderSimple', '_meta': {'title': 'Load Checkpoint'}}, '2': {'inputs': {'seed': 316381668905581, 'steps': 20, 'cfg': 8.0, 'sampler_name': 'euler', 'scheduler': 'normal', 'denoise': 1.0, 'model': ['8', 0], 'positive': ['3', 0], 'negative': ['4', 0], 'latent_image': ['6', 0]}, 'class_type': 'KSampler', '_meta': {'title': 'KSampler'}}, '3': {'inputs': {'text': 'An old man on a park bench', 'clip': ['8', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Prompt)'}}, '4': {'inputs': {'text': '', 'clip': ['8', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Prompt)'}}, '5': {'inputs': {'samples': ['2', 0], 'vae': ['1', 2]}, 'class_type': 'VAEDecode', '_meta': {'title': 'VAE Decode'}}, '6': {'inputs': {'width': 1024, 'height': 1024, 'batch_size': 1}, 'class_type': 'EmptyLatentImage', '_meta': {'title': 'Empty Latent Image'}}, '7': {'inputs': {'filename_prefix': 'ComfyUI', 'images': ['5', 0]}, 'class_type': 'SaveImage', '_meta': {'title': 'Save Image'}}, '8': {'inputs': {'lora_name': 'Hyper-SDXL-12steps-CFG-lora.safetensors', 'strength_model': 1.0, 'strength_clip': 1.0, 'model': ['1', 0], 'clip': ['1', 1]}, 'class_type': 'LoraLoader', '_meta': {'title': 'Load LoRA'}}, 3: {'inputs': {'text': 'An old man on a park bench', 'clip': ['8', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Prompt)'}}, 2: {'inputs': {'seed': 316381668905581, 'steps': 20, 'cfg': 8.0, 'sampler_name': 'euler', 'scheduler': 'normal', 'denoise': 1.0, 'model': ['8', 0], 'positive': ['3', 0], 'negative': ['4', 0], 'latent_image': ['6', 0]}, 'class_type': 'KSampler', '_meta': {'title': 'KSampler'}}}, name='Default Workflow')
# ---
# name: test_can_create_workflow_from_image
  Workflow(gui_workflow={'last_node_id': 8, 'last_link_id': 14, 'nodes': [{'id': 1, 'type': 'CheckpointLoaderSimple', 'pos': [593.9912109375, 1441.2381591796875], 'size': [315, 98], 'flags': {}, 'order': 1, 'mode': 0, 'inputs': [], 'outputs': [{'name': 'MODEL', 'type': 'MODEL', 'links': [10], 'slot_index': 0}, {'name': 'CLIP', 'type': 'CLIP', 'links': [12], 'slot_index': 1}, {'name': 'VAE', 'type': 'VAE', 'links': [6], 'slot_index': 2}], 'properties': {'Node name for S&R': 'CheckpointLoaderSimple'}, 'widgets_values': ['juggernautXL_juggernautX.safetensors']}, {'id': 2, 'type': 'KSampler', 'pos': [2061.75390625, 1429.724365234375], 'size': [315, 474], 'flags': {}, 'order': 5, 'mode': 0, 'inputs': [{'name': 'model', 'type': 'MODEL', 'link': 11}, {'name': 'positive', 'type': 'CONDITIONING', 'link': 2}, {'name': 'negative', 'type': 'CONDITIONING', 'link': 3}, {'name': 'latent_image', 'type': 'LATENT', 'link': 7}], 'outputs': [{'name': 'LATENT', 'type': 'LATENT', 'links': [8], 'slot_index': 0}], 'properties': {'Node name for S&R': 'KSampler'}, 'widgets_values': [176628554975370, 'randomize', 20, 8, 'euler', 'normal', 1]}, {'id': 3, 'type': 'CLIPTextEncode', 'pos': [1592.681396484375, 1540.5108642578125], 'size': [400, 200], 'flags': {}, 'order': 3, 'mode': 0, 'inputs': [{'name': 'clip', 'type': 'CLIP', 'link': 13}], 'outputs': [{'name': 'CONDITIONING', 'type': 'CONDITIONING', 'links': [2]}], 'properties': {'Node name for S&R': 'CLIPTextEncode'}, 'widgets_values': ['a magical purple orb']}, {'id': 4, 'type': 'CLIPTextEncode', 'pos': [1596.7044677734375, 1803.4591064453125], 'size': [400, 200], 'flags': {}, 'order': 4, 'mode': 0, 'inputs': [{'name': 'clip', 'type': 'CLIP', 'link': 14}], 'outputs': [{'name': 'CONDITIONING', 'type': 'CONDITIONING', 'links': [3]}], 'properties': {'Node name for S&R': 'CLIPTextEncode'}, 'widgets_values': ['']}, {'id': 5, 'type': 'VAEDecode', 'pos': [2473.87451171875, 1391.7969970703125], 'size': [210, 46], 'flags': {}, 'order': 6, 'mode': 0, 'inputs': [{'name': 'samples', 'type': 'LATENT', 'link': 8}, {'name': 'vae', 'type': 'VAE', 'link': 6}], 'outputs': [{'name': 'IMAGE', 'type': 'IMAGE', 'links': [9], 'slot_index': 0}], 'properties': {'Node name for S&R': 'VAEDecode'}}, {'id': 6, 'type': 'EmptyLatentImage', 'pos': [2057.55859375, 1739.7811279296875], 'size': [315, 106], 'flags': {}, 'order': 0, 'mode': 0, 'inputs': [], 'outputs': [{'name': 'LATENT', 'type': 'LATENT', 'links': [7]}], 'properties': {'Node name for S&R': 'EmptyLatentImage'}, 'widgets_values': [1024, 1024, 1]}, {'id': 7, 'type': 'SaveImage', 'pos': [2466.687744140625, 1545.7578125], 'size': [315, 270], 'flags': {}, 'order': 7, 'mode': 0, 'inputs': [{'name': 'images', 'type': 'IMAGE', 'link': 9}], 'outputs': [], 'properties': {}, 'widgets_values': ['ComfyUI']}, {'id': 8, 'type': 'LoraLoader', 'pos': [1196.7196044921875, 1665.0206298828125], 'size': [315, 126], 'flags': {}, 'order': 2, 'mode': 0, 'inputs': [{'name': 'model', 'type': 'MODEL', 'link': 10}, {'name': 'clip', 'type': 'CLIP', 'link': 12}], 'outputs': [{'name': 'MODEL', 'type': 'MODEL', 'links': [11], 'slot_index': 0}, {'name': 'CLIP', 'type': 'CLIP', 'links': [13, 14], 'slot_index': 1}], 'properties': {'Node name for S&R': 'LoraLoader'}, 'widgets_values': ['Hyper-SDXL-12steps-CFG-lora.safetensors', 1, 1]}], 'links': [[2, 3, 0, 2, 1, 'CONDITIONING'], [3, 4, 0, 2, 2, 'CONDITIONING'], [6, 1, 2, 5, 1, 'VAE'], [7, 6, 0, 2, 3, 'LATENT'], [8, 2, 0, 5, 0, 'LATENT'], [9, 5, 0, 7, 0, 'IMAGE'], [10, 1, 0, 8, 0, 'MODEL'], [11, 8, 0, 2, 0, 'MODEL'], [12, 1, 1, 8, 1, 'CLIP'], [13, 8, 1, 3, 0, 'CLIP'], [14, 8, 1, 4, 0, 'CLIP']], 'groups': [], 'config': {}, 'extra': {'ds': {'scale': 1.6940000000000053, 'offset': [-1357.8089102097, -1514.3590659107385]}, 'node_versions': {'comfy-core': '0.3.12'}, 'VHS_latentpreview': False, 'VHS_latentpreviewrate': 0}, 'version': 0.4}, api_workflow={'1': {'inputs': {'ckpt_name': 'juggernautXL_juggernautX.safetensors'}, 'class_type': 'CheckpointLoaderSimple', '_meta': {'title': 'Load Checkpoint'}}, '2': {'inputs': {'seed': 176628554975370, 'steps': 20, 'cfg': 8.0, 'sampler_name': 'euler', 'scheduler': 'normal', 'denoise': 1.0, 'model': ['8', 0], 'positive': ['3', 0], 'negative': ['4', 0], 'latent_image': ['6', 0]}, 'class_type': 'KSampler', '_meta': {'title': 'KSampler'}}, '3': {'inputs': {'text': 'a magical purple orb', 'clip': ['8', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Prompt)'}}, '4': {'inputs': {'text': '', 'clip': ['8', 1]}, 'class_type': 'CLIPTextEncode', '_meta': {'title': 'CLIP Text Encode (Prompt)'}}, '5': {'inputs': {'samples': ['2', 0], 'vae': ['1', 2]}, 'class_type': 'VAEDecode', '_meta': {'title': 'VAE Decode'}}, '6': {'inputs': {'width': 1024, 'height': 1024, 'batch_size': 1}, 'class_type': 'EmptyLatentImage', '_meta': {'title': 'Empty Latent Image'}}, '7': {'inputs': {'filename_prefix': 'ComfyUI', 'images': ['5', 0]}, 'class_type': 'SaveImage', '_meta': {'title': 'Save Image'}}, '8': {'inputs': {'lora_name': 'Hyper-SDXL-12steps-CFG-lora.safetensors', 'strength_model': 1.0, 'strength_clip': 1.0, 'model': ['1', 0], 'clip': ['1', 1]}, 'class_type': 'LoraLoader', '_meta': {'title': 'Load LoRA'}}}, name='Default Workflow')
# ---
# name: test_can_get_tweaks_from_file
  list([
    Tweak(selector={'id': '3'}, changes={'text': 'An old man on a park bench'}),
    Tweak(selector={'id': '2'}, changes={'seed': 387488229552385}),
  ])
# ---
# name: test_can_save_tweaks_to_yaml_file
  Tweaks(tweaks=[Tweak(selector={'id': '346'}, changes={'lora_name': 'test.safetensors'})], name='Default Tweaks', _original_yaml="tweaks:\n- changes:\n    lora_name: test.safetensors\n  selector:\n    id: '346'\n", _iteration=0)
# ---
